In [2]:
import numpy as np
import pandas as pd
import matplotlib as plt
import seaborn as sns
import plotly.express as px
from sklearn.preprocessing import StandardScaler, OneHotEncoder, FunctionTransformer

from sklearn.experimental import enable_hist_gradient_boosting
from sklearn.ensemble import HistGradientBoostingRegressor


In [3]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns

from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, OneHotEncoder, FunctionTransformer
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.metrics import plot_confusion_matrix, recall_score,\
    accuracy_score, precision_score, f1_score

from imblearn.over_sampling import SMOTE
from imblearn.pipeline import Pipeline as ImPipeline


from sklearn.experimental import enable_hist_gradient_boosting
from sklearn.ensemble import HistGradientBoostingRegressor
from sklearn.ensemble import HistGradientBoostingClassifier

In [4]:
X_train = pd.read_csv('../data/well_features.csv')
y_train = pd.read_csv('../data/well_labels.csv')
X_test = pd.read_csv('../data/well_test_labels.csv')

In [5]:
X_train.head()

id  amount_tsh date_recorded        funder  gps_height     installer  \
0  69572      6000.0    2011-03-14         Roman        1390         Roman   
1   8776         0.0    2013-03-06       Grumeti        1399       GRUMETI   
2  34310        25.0    2013-02-25  Lottery Club         686  World vision   
3  67743         0.0    2013-01-28        Unicef         263        UNICEF   
4  19728         0.0    2011-07-13   Action In A           0       Artisan   

   longitude   latitude              wpt_name  num_private  ... payment_type  \
0  34.938093  -9.856322                  none            0  ...     annually   
1  34.698766  -2.147466              Zahanati            0  ...    never pay   
2  37.460664  -3.821329           Kwa Mahundi            0  ...   per bucket   
3  38.486161 -11.155298  Zahanati Ya Nanyumbu            0  ...    never pay   
4  31.130847  -1.825359               Shuleni            0  ...    never pay   

  water_quality quality_group      quantity  quantity_group  \
0          soft          good        enough          enough   
1          soft          good  insufficient    insufficient   
2          soft          good        enough          enough   
3          soft          good           dry             dry   
4          soft          good      seasonal        seasonal   

                 source           source_type  source_class  \
0                spring                spring   groundwater   
1  rainwater harvesting  rainwater harvesting       surface   
2                   dam                   dam       surface   
3           machine dbh              borehole   groundwater   
4  rainwater harvesting  rainwater harvesting       surface   

               waterpoint_type waterpoint_type_group  
0           communal standpipe    communal standpipe  
1           communal standpipe    communal standpipe  
2  communal standpipe multiple    communal standpipe  
3  communal standpipe multiple    communal standpipe  
4           communal standpipe    communal standpipe  

[5 rows x 40 columns]

After Discussing and going over features, we decided on which ones to drop

In [6]:
dropped_list=['id','funder','wpt_name','num_private','basin','subvillage','region_code','district_code','lga','ward','recorded_by','scheme_name','extraction_type_group','extraction_type_class','management_group','payment_type','quality_group','quantity_group','source_type','source_class','waterpoint_type_group']

In [7]:
#Cleaning up the Datamframe
X_train = X_train.set_index(X_train['id'])
y_train = y_train.set_index(y_train['id'])
X_test = X_test.set_index(X_test['id'])
X_train = X_train.drop(dropped_list,axis=1)
X_test = X_test.drop(dropped_list,axis=1)
y_train = y_train.drop('id',axis=1)


In [8]:
y_train.head()

status_group
id                   
69572      functional
8776       functional
34310      functional
67743  non functional
19728      functional

In [9]:
X_test['date_recorded'] = X_test['date_recorded'].map(lambda x: x[:4]).astype('int64')
X_train['date_recorded'] = X_train['date_recorded'].map(lambda x: x[:4]).astype('int64')


In [11]:
X_train.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 59400 entries, 69572 to 26348
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   amount_tsh         59400 non-null  float64
 1   date_recorded      59400 non-null  int64  
 2   gps_height         59400 non-null  int64  
 3   installer          55745 non-null  object 
 4   longitude          59400 non-null  float64
 5   latitude           59400 non-null  float64
 6   region             59400 non-null  object 
 7   population         59400 non-null  int64  
 8   public_meeting     56066 non-null  object 
 9   scheme_management  55523 non-null  object 
 10  permit             56344 non-null  object 
 11  construction_year  59400 non-null  int64  
 12  extraction_type    59400 non-null  object 
 13  management         59400 non-null  object 
 14  payment            59400 non-null  object 
 15  water_quality      59400 non-null  object 
 16  quantity          

In [12]:
y_train.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 59400 entries, 69572 to 26348
Data columns (total 1 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   status_group  59400 non-null  object
dtypes: object(1)
memory usage: 928.1+ KB


In [12]:
y_train['status_group'].value_counts()

functional                 32259
non functional             22824
functional needs repair     4317
Name: status_group, dtype: int64

In [13]:
df_train = X_train.copy()
df_train = df_train.join(y_train, how='outer')

Creating our Pipeline and tranformers and grid search. Other members are running other models. Lets try the Histogradientclassfier model.

In [21]:
subpipe_num = Pipeline(steps=[('num_impute', SimpleImputer()),
                           ('ss', StandardScaler())])
subpipe_cat = Pipeline(steps=[('cat_impute', SimpleImputer(strategy='most_frequent')),
                             ('ohe', OneHotEncoder(sparse=False, handle_unknown='ignore'))])

In [22]:
#Pipeline Transformer
CT = ColumnTransformer(transformers=[('subpipe_num', subpipe_num, [0,1,2,4,5,7,11]),
                                         ('subpipe_cat', subpipe_cat, [3,6,8,9,10,12,13,14,15,16,17,18])],
                           remainder='passthrough')

In [23]:
pipe1 = Pipeline(steps = [('preprocessor', CT), ('classifier', HistGradientBoostingClassifier(random_state = 2))])

In [24]:
parameters = {'classifier__learning_rate': [.1, 1], 'classifier__max_iter': [100], 'classifier__max_depth': [10], 'classifier__l2_regularization': [0]}

In [26]:
grid = GridSearchCV(estimator= pipe1, param_grid = parameters, verbose = 1)
grid.fit(X_train, y_train)
grid.best_params_

Fitting 5 folds for each of 2 candidates, totalling 10 fits


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
C:\Users\01aut\anaconda3\envs\learn-env\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
C:\Users\01aut\anaconda3\envs\learn-env\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
C:\Users\01aut\anaconda3\envs\learn-env\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
C:\Users\01aut\anaconda3\envs\learn-env\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A col

{'classifier__l2_regularization': 0,
 'classifier__learning_rate': 0.1,
 'classifier__max_depth': 10,
 'classifier__max_iter': 100}

In [28]:
grid.cv_results_['mean_test_score']

array([0.78659933, 0.72107744])

We Got a cross validation score of 0.78659933 on the model with .1 learning rate and 0.72107744 on the model with a learning rate of 1. Lets change up these hyperparameters

In [29]:
parameters2 = {'classifier__learning_rate': [.1], 'classifier__max_iter': [100], 'classifier__max_depth': [10,15], 'classifier__l2_regularization': [0]}

In [30]:
grid2 = GridSearchCV(estimator= pipe1, param_grid = parameters2, verbose = 1)
grid2.fit(X_train, y_train)
grid2.best_params_

Fitting 5 folds for each of 2 candidates, totalling 10 fits


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
C:\Users\01aut\anaconda3\envs\learn-env\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
C:\Users\01aut\anaconda3\envs\learn-env\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
C:\Users\01aut\anaconda3\envs\learn-env\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
C:\Users\01aut\anaconda3\envs\learn-env\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A col

{'classifier__l2_regularization': 0,
 'classifier__learning_rate': 0.1,
 'classifier__max_depth': 15,
 'classifier__max_iter': 100}

In [31]:
grid2.cv_results_['mean_test_score']

array([0.78659933, 0.7906734 ])

Increasing the depth to 15 increase the score to 0.7906734. Lets continuing tuning the parameters.

In [34]:
parameters3 = {'classifier__learning_rate': [.1], 'classifier__max_iter': [100], 'classifier__max_depth': [15,20], 'classifier__l2_regularization': [0,2]}

In [36]:
grid3 = GridSearchCV(estimator= pipe1, param_grid = parameters3, verbose = 1)
grid3.fit(X_train, y_train)
grid3.best_params_

Fitting 5 folds for each of 4 candidates, totalling 20 fits


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
C:\Users\01aut\anaconda3\envs\learn-env\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
C:\Users\01aut\anaconda3\envs\learn-env\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
C:\Users\01aut\anaconda3\envs\learn-env\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
C:\Users\01aut\anaconda3\envs\learn-env\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A col

{'classifier__l2_regularization': 0,
 'classifier__learning_rate': 0.1,
 'classifier__max_depth': 20,
 'classifier__max_iter': 100}

In [37]:
grid3.cv_results_['mean_test_score']

array([0.7906734 , 0.79089226, 0.78824916, 0.78875421])

This time we got a score of 0.79089226 when increasing th depth to 20. However, changing out regularization penalty decrease our score at depth 15 and depth 20 to 0.78824916, 0.78875421 respectively.

In [39]:
parameters4 = {'classifier__learning_rate': [.1], 'classifier__max_iter': [100, 200], 'classifier__max_depth': [20,25], 'classifier__l2_regularization': [0]}

In [40]:
grid4 = GridSearchCV(estimator= pipe1, param_grid = parameters4, verbose = 1)
grid4.fit(X_train, y_train)
grid4.best_params_

Fitting 5 folds for each of 4 candidates, totalling 20 fits


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
C:\Users\01aut\anaconda3\envs\learn-env\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
C:\Users\01aut\anaconda3\envs\learn-env\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
C:\Users\01aut\anaconda3\envs\learn-env\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
C:\Users\01aut\anaconda3\envs\learn-env\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A col

{'classifier__l2_regularization': 0,
 'classifier__learning_rate': 0.1,
 'classifier__max_depth': 20,
 'classifier__max_iter': 200}

In [41]:
grid4.cv_results_['mean_test_score']

array([0.79089226, 0.79814815, 0.79040404, 0.79661616])

Further tuning increased our score to 0.79814815 at 200 iterations, but our score decreased when we changed our depth to 25 to 0.79040404 at 100 iterations and  0.79661616 at 200 iterations. Lets Try a normal GradientBoostingClassifier. 

In [42]:
pipe2 = Pipeline(steps = [('preprocessor', CT), ('classifier', GradientBoostingClassifier(random_state = 2))])

In [45]:
parameters5 = {'classifier__learning_rate': [.1], 'classifier__max_depth': [20]}

In [46]:
grid5 = GridSearchCV(estimator= pipe2, param_grid = parameters5, verbose = 1)
grid5.fit(X_train, y_train)
grid5.best_params_

Fitting 5 folds for each of 1 candidates, totalling 5 fits


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
C:\Users\01aut\anaconda3\envs\learn-env\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
C:\Users\01aut\anaconda3\envs\learn-env\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
C:\Users\01aut\anaconda3\envs\learn-env\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
C:\Users\01aut\anaconda3\envs\learn-env\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A col

{'classifier__learning_rate': 0.1, 'classifier__max_depth': 20}

In [47]:
grid5.cv_results_['mean_test_score']

array([0.80210438])

This was my best model at 0.80210438. Upon subbmission and putting this against test data, we had a score of 0.8086, which was better than our training data.